qiskit_optimization/algorithms/qrao/magic_rounding.py

```
@@ -0,0 +1,468 @@
+        for basis in bases:
+            if vars_per_qubit == 3:
+                qc = circuit.compose(_z_to_31p_qrac_basis_circuit(basis).inverse(), inplace=False)
+            elif vars_per_qubit == 2:
+                qc = circuit.compose(_z_to_21p_qrac_basis_circuit(basis).inverse(), inplace=False)
+            elif vars_per_qubit == 1:
+                qc = circuit.copy()
+            qc.measure_all()
+            circuits.append(qc)
+        return circuits
+
+    def _evaluate_magic_bases(
+        self,
+        circuit: QuantumCircuit,
+        bases: np.ndarray,
+        basis_shots: np.ndarray,
+        vars_per_qubit: int,
+    ) -> list[dict[str, int]]:
+        """
+        Given a quantum circuit to measure, a list of magic bases to measure, and a list of the
+        shots to use for each magic basis configuration, measure the provided circuit in the magic
+        bases given and return the counts dictionaries associated with each basis measurement.
+
+        Args:
+            circuit: The quantum circuit to measure.
+            bases: A list of magic bases to measure.
+            basis_shots: A list of shots to use for each magic basis configuration.
+            vars_per_qubit: The number of decision variables per qubit.
+
+        Returns:
+            A list of counts dictionaries associated with each basis measurement.
+
+        Raises:
+            AlgorithmError: If the primitive job failed.
+        """
+        circuits = self._make_circuits(circuit, bases, vars_per_qubit)
+        # Execute each of the rotated circuits and collect the results
+        # Batch the circuits into jobs where each group has the same number of
+        # shots, so that you can wait for the queue as few times as possible if
+        # using hardware.
+        circuit_indices_by_shots: dict[int, list[int]] = defaultdict(list)
+        basis_counts: list[dict[str, int] | None] = [None] * len(circuits)
+        assert len(circuits) == len(basis_shots)
```

`assert` should not be used. Need to raise an appropriate error or a warning.